README.md
# kube-lego

*kube-lego* automatically requests certificates for Kubernetes Ingress resources from Let's Encrypt

[![Build Status](https://travis-ci.org/jetstack/kube-lego.svg?branch=master)](https://travis-ci.org/jetstack/kube-lego)
[![](https://images.microbadger.com/badges/version/jetstack/kube-lego.svg)](http://microbadger.com/#/images/jetstack/kube-lego "Get your own version badge on microbadger.com")

## Screencast

[![Kube Lego screencast](https://asciinema.org/a/47444.png)](https://asciinema.org/a/47444)

## Features

- Recognizes the need of a new certificate for this cases:
  - No certificate existing
  - Existing certificate is not containing all domain names
  - Existing certificate is expired or near to its expiry date (cf. option `LEGO_MINIMUM_VALIDITY`)
  - Existing certificate is unparseable, invalid or not matching the secret key
- Creates a user account (incl. private key) for Let's Encrypt and stores it in Kubernetes secrets (secret name is configurable via `LEGO_SECRET_NAME`)
- Obtains the missing certificates from Let's Encrypt and authorizes the request with the `HTTP-01` challenge
- Makes sure that the specific Kubernetes objects (Services, Ingress) contain the rights configuration for the `HTTP-01` challenge to succeed
- Official Kubernetes Helm [chart](https://github.com/kubernetes/charts/tree/master/stable/kube-lego) for simplistic deployment.

## Requirements

- Kubernetes 1.2+
- Compatible ingress controller (nginx or GCE see [here](#ingress-controllers))
- Non-production use case :laughing:

## Usage

### run kube-lego

* [GCE](examples/gce/README.md)
* [nginx controller](examples/nginx/README.md)

The default value of `LEGO_URL` is the Let's Encrypt **staging environment**. If you want to get "real" certificates you have to configure their production env.

If you change the `LEGO_URL`, it is required that you delete the existing secret `kube-lego-account` and all certificates you want to request from the new URL.

### how kube-lego works

As soon as the kube-lego daemon is running, it will create a user account with LetsEncrypt, make a service resource, and look for ingress resources that have this annotation:

```yaml
metadata:
  annotations:
    kubernetes.io/tls-acme: "true"
```

Every ingress resource that has this annotation will be monitored by *kube-lego* (cluster-wide in all namespaces). The only part that is watched is the list `spec.tls`. Every element will get its own certificate through Let's Encrypt.

Let's take a look at this ingress resource:

```yaml
spec:
  tls:
  - secretName: mysql-tls
    hosts:
    - phpmyadmin.example.com
    - mysql.example.com
  - secretName: postgres-tls
    hosts:
    - postgres.example.com
```

On finding the above resource, the following happens:

1. An ingress resource is created coordinating where to send acme challenges for the said domains.

2. *kube-lego* will then perform its own check for i.e. `http://mysql.example.com/.well-known/acme-challenge/_selftest` to ensure all is well before reaching out to letsencrypt.

3. *kube-lego* will obtain two certificates (one with phpmyadmin.example.com and mysql.example.com, the other with postgres.example.com).


Please note:

- The `secretName` statements have to be unique per namespace
- `secretName` is required (even if no secret exists with that name, as it will be created by *kube-lego*)
- Setups which utilize 1:1 NAT need to ensure internal resources can reach gateway controlled public addresses.
- Additionally, your domain must point to your externally available Load Balancer (either directly or via 1:1 NAT)


## Ingress controllers

### [Nginx Ingress Controller](https://github.com/kubernetes/ingress/tree/master/controllers/nginx)

- available through image `gcr.io/google_containers/nginx-ingress-controller`
- fully supports kube-lego from version 0.8 onwards

### [GCE Loadbalancers](https://github.com/kubernetes/ingress/tree/master/controllers/gce)

- you don't have to maintain the ingress controller yourself, you pay GCE to do that for you
- every ingress resource creates one GCE load balancer
- all service that you want to expose, have to be `Type=NodePort`

## Environment variables

| Name | Required | Default | Description |
|------|----------|---------|-------------|
| `LEGO_EMAIL` | y | `-` | E-Mail address for the ACME account, used to recover from lost secrets |
| `LEGO_POD_IP` | y | `-` | Pod IP address (use the [downward API](https://kubernetes.io/docs/tasks/configure-pod-container/environment-variable-expose-pod-information/#the-downward-api))|
| `LEGO_NAMESPACE` | n | `default` | Namespace where kube-lego is running in |
| `LEGO_URL` | n | `https://acme-staging.api.letsencrypt.org/directory` | URL for the ACME server. To get "real" certificates set to the production API of Let's Encrypt: `https://acme-v01.api.letsencrypt.org/directory` |
| `LEGO_SECRET_NAME` | n | `kube-lego-account` | Name of the secret in the same namespace that contains ACME account secret |
| `LEGO_SERVICE_NAME_NGINX` | n | `kube-lego-nginx` | Service name for NGINX ingress |
| `LEGO_SERVICE_NAME_GCE` | n | `kube-lego-gce` | Service name for GCE ingress |
| `LEGO_SUPPORTED_INGRESS_CLASS` | n | `nginx,gce` | Specify the supported ingress class |
| `LEGO_SUPPORTED_INGRESS_PROVIDER` | n | `nginx,gce` | Specify the supported ingress provider |
| `LEGO_INGRESS_NAME_NGINX` | n | `kube-lego-nginx` | Ingress name which contains the routing for HTTP verification for nginx ingress |
| `LEGO_PORT` | n | `8080` | Port where this daemon is listening for verifcation calls (HTTP method)|
| `LEGO_CHECK_INTERVAL` | n | `8h` | Interval for periodically certificate checks (to find expired certs)|
| `LEGO_MINIMUM_VALIDITY` | n | `720h` (30 days) | Request a renewal when the remaining certificate validity falls below that value|
| `LEGO_DEFAULT_INGRESS_CLASS` | n | `nginx` | Default ingress class for resources without specification|
| `LEGO_KUBE_API_URL` | n | `http://127.0.0.1:8080` | API server URL |
| `LEGO_LOG_LEVEL` | n | `info` | Set log level (`debug`, `info`, `warn` or `error`) |
| `LEGO_LOG_TYPE` | n | `text` | Set log type. Only `json` as custom value supported, everything else defaults to default logrus textFormat |
| `LEGO_KUBE_ANNOTATION` | n | `kubernetes.io/tls-acme` | Set the ingress annotation used by this instance of kube-lego to get certificate for from Let's Encrypt. Allows you to run kube-lego against staging and production LE |
| `LEGO_WATCH_NAMESPACE` | n | `` | Namespace that kube-lego should watch for ingresses and services |

## Full deployment examples

- [Nginx Ingress Controller](examples/nginx/)
- [GCE Load Balancers](examples/gce/)

## Troubleshooting

When interacting with *kube-lego*, its a good idea to run with `LEGO_LOG_LEVEL=debug` for more verbose details.
Additionally, be aware of the automatically created resources (see environment variables) when cleaning up or testing.

Possible resources for *help*:

* The official channel `#kube-lego` on `kubernetes.slack.com`

> There is also a good chance to get some support on non-official support
> channels for *kube-lego*, but be aware that these are rather general
> kubernetes discussion channels.

* `#coreos` on freenode
* Slack channels like `#kubernetes-users` or `#kubernetes-novice` on `kubernetes.slack.com`
* If you absolutely just can't figure out your problem, file an issue.


### Enable the pprof tool

To enable the [pprof tool](https://golang.org/pkg/net/http/pprof/) run kube-lego with environment `LEGO_LOG_LEVEL=debug`.

Capture 20 seconds of the execution trace:

`$ wget http://localhost:8080/debug/pprof/trace?seconds=20 -O kube-lego.trace`

You can inspect the trace sample running 

`$ go tool trace kube-lego.trace`


## Authors

* Christian Simon for [Jetstack Ltd](http://www.jetstack.io)
* [PR contributors](https://github.com/jetstack/kube-lego/graphs/contributors)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.5] - 2017-07-03

### Fixed

- Fix infinite update loop on ingress changes [#204](https://github.com/jetstack/kube-lego/pull/204)
- Fix error message wording [#209](https://github.com/jetstack/kube-lego/pull/209)
- Fix documentation wording [#208](https://github.com/jetstack/kube-lego/pull/208)

### Added

- Whitelist ingress IP source range (allow 0.0.0.0/0) [#114](https://github.com/jetstack/kube-lego/pull/114)

## [0.1.4] - 2017-05-12

### Fixed

- Set the ENTRYPOINT instead of the CMD [#117](https://github.com/jetstack/kube-lego/pull/117)
- Fix suppressed error while ACME client creation [#155](https://github.com/jetstack/kube-lego/pull/155)
- Fix the order of namespace/name in the secret names [#141](https://github.com/jetstack/kube-lego/pull/141)
- fix logic typo #110

### Added

* Add support for pprof [#139](https://github.com/jetstack/kube-lego/pull/139)
* Update godeps and k8s method signatures [#48](https://github.com/jetstack/kube-lego/pull/48)
* Upgrade docker golang to v1.8 [#127](https://github.com/jetstack/kube-lego/pull/127)
* Add support for running multiple instance of kube-lego in the cluster [#105](https://github.com/jetstack/kube-lego/pull/105)
* Add ability to customize ingress class [#101](https://github.com/jetstack/kube-lego/pull/101)
* Change docker image from scratch to alpine [#87](https://github.com/jetstack/kube-lego/pull/87)
* Add ability to only monitor a single namespace [#123](https://github.com/jetstack/kube-lego/pull/123)

[View](https://github.com/jetstack/kube-lego/pulls?utf8=✓&q=is%3Apr%20is%3Aclosed%20merged%3A2016-11-03..2017-05-12%20%20fix) all the changes

## [0.1.3] - 2016-11-03

### Fixed

- Workaround for GLBC health check detection bug
- Raise resync period to 60 seconds
- Use correct minimum validity config (used the `LEGO_CHECK_INTERVAL` before)

### Added

* Replace lego with the acme library
* Increase tests coverage and automation
* Improved log output, no secrets default log level is debug
* Upgrade dependencies
* Upgrade to go 1.7.3
* Possibility to change api server url [Pavel Sorejs]

## [0.1.2] - 2016-08-19

### Fixed

- Better error output for failing validations
- Fix problems with updating ingress objects

## [0.1.1] - 2016-08-16

### Fixed

- Fix version output of kube-lego
- Fix image version in the docs

## [0.1.0] - 2016-08-10

### Added

- Support for GCE load balancer ingress controller
- E2E automation scripts for GCE/NGINX ingress controllers on GKE
- Support for ingress-class annotations to distinguish between GCE/NGINX ingress
- Abstracted the ingress controller specific code into separate packages
- Deployment of kube-lego uses readiness checks

### Fixed

- Handle failed certificate request without exiting kube-lego


## [0.0.4] - 2016-07-11
### Added
- Check for expired certificates periodically (default config every 8 hours)
- Use upstream nginx-ingress-controller (from k8s-contrib)

### Fixed
- Fix bug for empty kube-lego ingress resource

## [0.0.3] - 2016-05-27
### Added
- E2E test for receiving a cert from Let's Encrypt Staging
- Updating docs/examples to use latest nginx-ingress release

## [0.0.2] - 2016-05-03
### Added
- Documentation
- Versioned docker images
- Jenkins builds using pipelines/workflow plugin

## [0.0.1] - 2016-05-03 - MVP release
### Added
- Recognizes the need of a new certificate (domain name missing, certificate expired, certificate unparseable)
- Obtains certificates per TLS object in ingress resources and stores it in Kubernetes secrets using `HTTP-01` challenge
- Creates a user account (incl. private key) for Let's Encrypt and stores it in Kubernetes secrets
- Watches changes of ingress resources and reevaluate certificates
- Configures endpoints for `HTTP-01` challenge in a separate ingress resource


pkg/kubelego/kubelego.go
package kubelego

import (
	"errors"
	"fmt"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/jetstack/kube-lego/pkg/acme"
	"github.com/jetstack/kube-lego/pkg/ingress"
	kubelego "github.com/jetstack/kube-lego/pkg/kubelego_const"
	"github.com/jetstack/kube-lego/pkg/provider/gce"
	"github.com/jetstack/kube-lego/pkg/provider/nginx"
	"github.com/jetstack/kube-lego/pkg/secret"

	log "github.com/Sirupsen/logrus"
	"k8s.io/apimachinery/pkg/util/intstr"
	"k8s.io/client-go/kubernetes"
	k8sApi "k8s.io/client-go/pkg/api/v1"
)

var _ kubelego.KubeLego = &KubeLego{}

func makeLog() *log.Entry {
	logtype := strings.ToLower(os.Getenv("LEGO_LOG_TYPE"))
	if logtype == "" {
		logtype = "text"
	}

	if logtype == "json" {
		log.SetFormatter(&log.JSONFormatter{})
	} else if logtype == "text" {
		log.SetFormatter(&log.TextFormatter{})
	} else {
		log.WithField("logtype", logtype).Fatal("Given logtype was not valid, check LEGO_LOG_TYPE configuration")
		os.Exit(1)
	}

	loglevel := strings.ToLower(os.Getenv("LEGO_LOG_LEVEL"))
	if len(loglevel) == 0 {
		log.SetLevel(log.InfoLevel)
	} else if loglevel == "debug" {
		log.SetLevel(log.DebugLevel)
	} else if loglevel == "info" {
		log.SetLevel(log.InfoLevel)
	} else if loglevel == "warn" {
		log.SetLevel(log.WarnLevel)
	} else if loglevel == "error" {
		log.SetLevel(log.ErrorLevel)
	} else {
		log.SetLevel(log.InfoLevel)
	}
	return log.WithField("context", "kubelego")
}

func New(version string) *KubeLego {
	return &KubeLego{
		version:   version,
		log:       makeLog(),
		stopCh:    make(chan struct{}),
		waitGroup: sync.WaitGroup{},
	}
}

func (kl *KubeLego) Log() *log.Entry {
	return kl.log
}

func (kl *KubeLego) Stop() {
	kl.Log().Info("shutting things down")
	close(kl.stopCh)
}

func (kl *KubeLego) IngressProvider(name string) (provider kubelego.IngressProvider, err error) {
	provider, ok := kl.legoIngressProvider[name]
	if !ok {
		return nil, fmt.Errorf("Ingress provider '%s' not found", name)
	}
	return
}

func (kl *KubeLego) Init() {
	kl.Log().Infof("kube-lego %s starting", kl.version)

	// handle sigterm correctly
	k := make(chan os.Signal, 1)
	signal.Notify(k, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		s := <-k
		logger := kl.Log().WithField("signal", s.String())
		logger.Debug("received signal")
		kl.Stop()
	}()

	// parse env vars
	err := kl.paramsLego()
	if err != nil {
		kl.Log().Fatal(err)
	}

	kl.legoIngressProvider = map[string]kubelego.IngressProvider{}
	for _, provider := range kl.legoSupportedIngressProvider {
		switch provider {
		case "gce":
			kl.legoIngressProvider["gce"] = gce.New(kl)
			break
		case "nginx":
			kl.legoIngressProvider["nginx"] = nginx.New(kl)
			break
		default:
			kl.Log().Warnf("Unsupported provider [%s], please add a handler in kubelego.go#Init()", provider)
			break
		}
	}

	// start workers
	kl.WatchReconfigure()

	// intialize kube api
	err = kl.InitKube()
	if err != nil {
		kl.Log().Fatal(err)
	}

	// run acme http server
	myAcme := acme.New(kl)
	go func() {
		kl.waitGroup.Add(1)
		defer kl.waitGroup.Done()
		myAcme.RunServer(kl.stopCh)
	}()
	kl.acmeClient = myAcme

	// run ticker to check certificates periodically
	ticker := time.NewTicker(kl.legoCheckInterval)
	go func() {
		for timestamp := range ticker.C {
			kl.Log().Infof("Periodically check certificates at %s", timestamp)
			kl.requestReconfigure()
		}
	}()

	// watch for ingress controller events
	kl.WatchEvents()

	// wait for stop signal
	<-kl.stopCh
	ticker.Stop()
	kl.Log().Infof("exiting")
	kl.waitGroup.Wait()
}

func (kl *KubeLego) AcmeClient() kubelego.Acme {
	return kl.acmeClient
}

func (kl *KubeLego) KubeClient() *kubernetes.Clientset {
	return kl.kubeClient
}

func (kl *KubeLego) Version() string {
	return kl.version
}

func (kl *KubeLego) LegoHTTPPort() intstr.IntOrString {
	return kl.legoHTTPPort
}

func (kl *KubeLego) LegoURL() string {
	return kl.legoURL
}

func (kl *KubeLego) LegoEmail() string {
	return kl.legoEmail
}

func (kl *KubeLego) LegoNamespace() string {
	return kl.legoNamespace
}

func (kl *KubeLego) LegoWatchNamespace() string {
	return kl.legoWatchNamespace
}

func (kl *KubeLego) LegoPodIP() net.IP {
	return kl.legoPodIP
}

func (kl *KubeLego) LegoDefaultIngressClass() string {
	return kl.legoDefaultIngressClass
}

func (kl *KubeLego) LegoIngressNameNginx() string {
	return kl.legoIngressNameNginx
}
func (kl *KubeLego) LegoSupportedIngressClass() []string {
	return kl.legoSupportedIngressClass
}

func (kl *KubeLego) LegoSupportedIngressProvider() []string {
	return kl.legoSupportedIngressProvider
}

func (kl *KubeLego) LegoServiceNameNginx() string {
	return kl.legoServiceNameNginx
}

func (kl *KubeLego) LegoServiceNameGce() string {
	return kl.legoServiceNameGce
}

func (kl *KubeLego) LegoMinimumValidity() time.Duration {
	return kl.legoMinimumValidity
}

func (kl *KubeLego) LegoCheckInterval() time.Duration {
	return kl.legoCheckInterval
}

func (kl *KubeLego) LegoKubeApiURL() string {
	return kl.legoKubeApiURL
}

func (kl *KubeLego) acmeSecret() *secret.Secret {
	return secret.New(kl, kl.LegoNamespace(), kl.legoSecretName)
}

func (kl *KubeLego) AcmeUser() (map[string][]byte, error) {
	s := kl.acmeSecret()
	if !s.Exists() {
		return map[string][]byte{}, fmt.Errorf("no acme user found %s/%s", kl.LegoNamespace(), kl.legoSecretName)
	}
	return s.SecretApi.Data, nil
}

func (kl *KubeLego) SaveAcmeUser(data map[string][]byte) error {
	s := kl.acmeSecret()
	s.SecretApi.Data = data
	return s.Save()
}

// read config parameters from ENV vars
func (kl *KubeLego) paramsLego() error {

	kl.legoEmail = os.Getenv("LEGO_EMAIL")
	if len(kl.legoEmail) == 0 {
		return errors.New("Please provide an email address for certificate expiration notifications in LEGO_EMAIL (https://letsencrypt.org/docs/expiration-emails/)")
	}

	kl.legoPodIP = net.ParseIP(os.Getenv("LEGO_POD_IP"))
	if kl.legoPodIP == nil {
		return errors.New("Please provide the pod's IP via environment variable LEGO_POD_IP using the downward API (http://kubernetes.io/docs/user-guide/downward-api/)")
	}

	kl.legoNamespace = os.Getenv("LEGO_NAMESPACE")
	if len(kl.legoNamespace) == 0 {
		kl.legoNamespace = k8sApi.NamespaceDefault
	}

	kl.legoURL = os.Getenv("LEGO_URL")
	if len(kl.legoURL) == 0 {
		kl.legoURL = "https://acme-staging.api.letsencrypt.org/directory"
	}

	kl.legoSecretName = os.Getenv("LEGO_SECRET_NAME")
	if len(kl.legoSecretName) == 0 {
		kl.legoSecretName = "kube-lego-account"
	}

	kl.legoServiceNameNginx = os.Getenv("LEGO_SERVICE_NAME_NGINX")
	if len(kl.legoServiceNameNginx) == 0 {
		kl.legoServiceNameNginx = os.Getenv("LEGO_SERVICE_NAME")
		if len(kl.legoServiceNameNginx) == 0 {
			kl.legoServiceNameNginx = "kube-lego-nginx"
		}
	}

	kl.legoServiceNameGce = os.Getenv("LEGO_SERVICE_NAME_GCE")
	if len(kl.legoServiceNameGce) == 0 {
		kl.legoServiceNameGce = "kube-lego-gce"
	}

	supportedProviders := os.Getenv("LEGO_SUPPORTED_INGRESS_PROVIDER")
	if len(supportedProviders) == 0 {
		kl.legoSupportedIngressProvider = kubelego.SupportedIngressProviders
	} else {
		kl.legoSupportedIngressProvider = strings.Split(supportedProviders, ",")
	}

	legoSupportedIngressClass := os.Getenv("LEGO_SUPPORTED_INGRESS_CLASS")
	if len(legoSupportedIngressClass) == 0 {
		kl.legoSupportedIngressClass = kubelego.SupportedIngressClasses
	} else {
		kl.legoSupportedIngressClass = strings.Split(legoSupportedIngressClass, ",")
	}

	legoDefaultIngressClass := os.Getenv("LEGO_DEFAULT_INGRESS_CLASS")
	if len(legoDefaultIngressClass) == 0 {
		kl.legoDefaultIngressClass = "nginx"
	} else {
		var err error = nil
		kl.legoDefaultIngressClass, err = ingress.IsSupportedIngressClass(kl.legoSupportedIngressClass, legoDefaultIngressClass)
		if err != nil {
			return fmt.Errorf("Unsupported default ingress class: '%s'. You can set the ingress class with 'LEGO_DEFAULT_INGRESS_CLASS'", legoDefaultIngressClass)
		}
	}
	kl.legoIngressNameNginx = os.Getenv("LEGO_INGRESS_NAME_NGINX")
	if len(kl.legoIngressNameNginx) == 0 {
		kl.legoIngressNameNginx = os.Getenv("LEGO_INGRESS_NAME")
		if len(kl.legoIngressNameNginx) == 0 {
			kl.legoIngressNameNginx = "kube-lego-nginx"
		}
	}

	checkIntervalString := os.Getenv("LEGO_CHECK_INTERVAL")
	if len(checkIntervalString) == 0 {
		kl.legoCheckInterval = 8 * time.Hour
	} else {
		d, err := time.ParseDuration(checkIntervalString)
		if err != nil {
			return err
		}
		if d < 5*time.Minute {
			return fmt.Errorf("Minimum check interval is 5 minutes: %s", d)
		}
		kl.legoCheckInterval = d
	}

	kl.legoKubeApiURL = os.Getenv("LEGO_KUBE_API_URL")
	if len(kl.legoKubeApiURL) == 0 {
		kl.legoKubeApiURL = "http://127.0.0.1:8080"
	}

	minimumValidity := os.Getenv("LEGO_MINIMUM_VALIDITY")
	if len(minimumValidity) == 0 {
		kl.legoMinimumValidity = time.Hour * 24 * 30
	} else {
		d, err := time.ParseDuration(minimumValidity)
		if err != nil {
			return err
		}
		if d < 24*time.Hour {
			return fmt.Errorf("Smallest allowed minimum validity is 24 hours: %s", d)
		}
		kl.legoMinimumValidity = d
	}

	httpPortStr := os.Getenv("LEGO_PORT")
	if len(httpPortStr) == 0 {
		kl.legoHTTPPort = intstr.FromInt(8080)
	} else {
		i, err := strconv.Atoi(httpPortStr)
		if err != nil {
			return err
		}
		if i <= 0 || i >= 65535 {
			return fmt.Errorf("Wrong port: %d", i)
		}
		kl.legoHTTPPort = intstr.FromInt(i)
	}

	annotationEnabled := os.Getenv("LEGO_KUBE_ANNOTATION")
	if len(annotationEnabled) != 0 {
		kubelego.AnnotationEnabled = annotationEnabled
	}

	watchNamespace := os.Getenv("LEGO_WATCH_NAMESPACE")
	if len(watchNamespace) == 0 {
		kl.legoWatchNamespace = k8sApi.NamespaceAll
	} else {
		kl.legoWatchNamespace = watchNamespace
	}
	return nil
}
